### PR TITLE
Added test reporters and coverage analysis

### PR DESCRIPTION
**Pull request by Slartibartfass2**

### CODE REVIEW

File: .github/semantic.yml
Comment (Merseleo) [inline .github/semantic.yml:1]: Only for my personal interest: what is this?

File: src/app.css
Comment (Merseleo) [inline src/app.css:1]: Is this because no rebase was done so far? (seems to be same for some other files, like the src/lib/utils, src/routes/theme/...

File: .gitignore
Comment (Merseleo) [inline .gitignore:1]: Maybe I should add this update to the other PR #78 and then you can rebase on develop and include the updates? Or do you think it is better to add them here?

File: e2e/demo.test.ts
Comment (Merseleo) [inline e2e/demo.test.ts:1]: Maybe should rename this folder? Suggestion: e2e_tests/

File: vite.config.ts [hunk 1]
```diff
@@ -1,10 +1,24 @@
...
 
 export default defineConfig({
-    plugins: [sveltekit()],
-
+    plugins: [sveltekit(), svelteTesting()],
+    // Tell Vitest to use the `browser` entry points in `package.json` files, even though it's running in Node
+    resolve: process.env.VITEST ? { conditions: ["browser"] } : undefined,
     test: {
-        include: ["test/**/*.{test,spec}.{js,ts}"],
+        include: ["test/**/*.{test,spec}.{js,ts}", "integration/**/*.{test,spec}.{js,ts}"],
```
Comment (Merseleo) [inline vite.config.ts]: Do we have an integration directory? For components testing?

File: .github/workflows/code_quality_checks.yml
Comment (Merseleo) [inline .github/workflows/code_quality_checks.yml:1]: In your config you added something regarding integration tests, do we want an empty job for the integration test already?

File: .github/workflows/code_quality_checks.yml [hunk 2]
```diff
@@ -48,6 +51,17 @@ jobs:
...
+
+            - name: Report code coverage
+              uses: zgosalvez/github-actions-report-lcov@v3
+              with:
+                  coverage-files: coverage/lcov.info
+                  artifact-name: code-coverage-report
+                  github-token: ${{ secrets.GITHUB_TOKEN }}
+                  update-comment: true
+
     e2e-test:
```
Comment (Merseleo) [inline .github/workflows/code_quality_checks.yml:87]: Maybe add to this job the upload of the playwright report (https://playwright.dev/docs/ci).

**Review by Merseleo — COMMENTED** [review]
Again, I think this is a good basis for all the testing and further configuration will follow when writing the tests. However, maybe you could add a meaningful test example, if this is possible (i. e. for a e2e test and a component test, which is probably not possible so far as we don't have any components completed, and normal tests). 
And as far as I understood examples etc. correctly the configs look good for now.


One last request: maybe you should add a short section about testing in the Readme.md, so how we test, how to run them locally and what is reported in the CI (optionally). Or do you think differently?

File: .github/semantic.yml
Comment (Slartibartfass2) [reply .github/semantic.yml:1]: This is a config file for the Semantic PRs GitHub App I added, which is like a workflow and checks whether the commits in a PR adhere to the conventional commits schema.

File: vite.config.ts (hunk 1 above)
Comment (Slartibartfass2) [reply vite.config.ts]: Currently, we don't because we have no tests but as soon as we add them, e.g. in #74 this directory will also be tracked in git.

File: .github/workflows/code_quality_checks.yml
Comment (Slartibartfass2) [reply .github/workflows/code_quality_checks.yml:1]: Both unit and integration tests are executed by vites with the `test:unit` command (I see that this name might be misleading).

File: .github/workflows/code_quality_checks.yml (hunk 2 above)
Comment (Slartibartfass2) [reply .github/workflows/code_quality_checks.yml:87]: Good catch, I added it.

File: src/app.css
Comment (Slartibartfass2) [reply src/app.css:1]: Since this PR has been open for some while and I replaced the default branch main with develop, this PR had main still as base, I changed that now, and this isn't part of the changes anymore.

File: e2e/demo.test.ts
Comment (Slartibartfass2) [reply e2e/demo.test.ts:1]: What do you think about the following structure?
```
test/
├─ e2e/
├─ integration/
└─ unit/
```

File: .gitignore
Comment (Slartibartfass2) [reply .gitignore:1]: See my comment in your PR: https://github.com/SE-UUlm/snowballr-frontend/pull/78#discussion_r1844978171

**Review by Slartibartfass2 — COMMENTED** [review]
I'll add a real component test in #74, I don't think it makes much sense to add a demo test now.

I'll add a section in the README as soon as we are finished discussing the location of the tests.

File: .github/workflows/code_quality_checks.yml
Comment (Slartibartfass2) [reply .github/workflows/code_quality_checks.yml:1]: I'm currently looking into the [workspace feature](https://vitest.dev/guide/workspace.html) of vitest, which could solve this.

We could have two configs for unit and integration tests and run them separately.

File: .github/semantic.yml
Comment (Merseleo) [reply .github/semantic.yml:1]: Ah perfect.

File: src/app.css
Comment (Merseleo) [reply src/app.css:1]: Thank you, it more clearly now.

File: .gitignore
Comment (Merseleo) [reply .gitignore:1]: I saw and updated it. As mine is probably merged before, I think this is fine and further updates of the .gitignore can be added from the other PRs (like the playwright report).

File: e2e/demo.test.ts
Comment (Merseleo) [reply e2e/demo.test.ts:1]: Afaik Svelte suggests to add the unit tests in the folders of the corresponding source files, but imo this structure is the best solution!

If you want to ask @luca-schlecker too or have any other idea, lets discuss it, otherwise resolve this thread on your own.

File: vite.config.ts [hunk 3]
```diff
@@ -1,10 +1,22 @@
...
+    // Tell Vitest to use the `browser` entry points in `package.json` files, even though it's running in Node
+    resolve: process.env.VITEST ? { conditions: ["browser"] } : undefined,
     test: {
-        include: ["test/**/*.{test,spec}.{js,ts}"],
+        reporters: process.env.GITHUB_ACTIONS ? ["default", "github-actions"] : ["default"],
+        sequence: {
+            shuffle: true,
+        },
+        coverage: {
+            include: ["src/**"],
```
Comment (Merseleo) [inline vite.config.ts:15]: Why now include src/** and not test/**?

File: .github/workflows/code_quality_checks.yml (hunk 2 above)
Comment (Merseleo) [reply .github/workflows/code_quality_checks.yml:87]: Perfect.

File: .github/workflows/code_quality_checks.yml
Comment (Merseleo) [reply .github/workflows/code_quality_checks.yml:1]: Yes, I think you find a very good solution. Imo the config files are also separated better now and the names are not misleading anymore. Perfect, thank you!

File: .github/workflows/code_quality_checks.yml [hunk 4]
```diff
@@ -68,3 +105,81 @@ jobs:
...
+
+            - name: Report code coverage
+              uses: zgosalvez/github-actions-report-lcov@v4
+              with:
+                  coverage-files: final-lcov.info
+                  artifact-name: code-coverage-report
+                  github-token: ${{ secrets.GITHUB_TOKEN }}
+                  update-comment: true
+
+    check:
```
Comment (Merseleo) [inline .github/workflows/code_quality_checks.yml:169]: What is the "check" job?

File: .github/workflows/code_quality_checks.yml (hunk 2 above)
Comment (Merseleo) [reply .github/workflows/code_quality_checks.yml:87]: Perfect, thank you.

File: vite.config.ts (hunk 3 above)
Comment (Slartibartfass2) [reply vite.config.ts:15]: Here we define which files are tracked for the coverage report. Without this, some config files in the root directory will be included.

File: e2e/demo.test.ts
Comment (Slartibartfass2) [reply e2e/demo.test.ts:1]: Afaik, we already discussed this and decided to have tests separately, which is why I'll resolve this thread.

**Review by Merseleo — DISMISSED** [review]
I think we are on the right track.
The new section in the README.md surely is helpful for new people and for us at beginning too. But one last question (must not be done in this PR): do you think it is helpful to host the report anywhere and link to them in the README.md or add a coverage badge or something like that?

File: vite.config.ts (hunk 3 above)
Comment (Merseleo) [reply vite.config.ts:15]: Ah ok, makes sense

File: .github/workflows/code_quality_checks.yml (hunk 4 above)
Comment (Slartibartfass2) [reply .github/workflows/code_quality_checks.yml:169]: You can read about it [here](https://svelte.dev/docs/cli/sv-check).

TL;DR it mainly checks if the TypeScript code compiles.

**Comment by Slartibartfass2** [comment]
> I think we are on the right track. The new section in the README.md surely is helpful for new people and for us at beginning too. But one last question (must not be done in this PR): do you think it is helpful to host the report anywhere and link to them in the README.md or add a coverage badge or something like that?

we could use github pages to host, but I am not convinced that's necessary. If you want to check out the coverage, one could either download the report or just run the tests in the repo.

About the coverage badge: Maybe that's possible, but I'm not sure. Perhaps we can research this and add this later.

File: .github/workflows/code_quality_checks.yml (hunk 4 above)
Comment (Merseleo) [reply .github/workflows/code_quality_checks.yml:169]: Ah ok, seems to be useful, good idea

**Review by Merseleo — APPROVED** [review]
LGTM
